2023-07-28-on-demand-builders.md

```diff
@@ -0,0 +1,125 @@
+On demand started builders in Copr
+==================================
+
+The problem is that our builders are too slow to build e.g. Chromium package
+quickly enough.  Such builds take 24h and more (our limit is 48h).  With the
+cloud era, it really makes sense to have more powerful machines -> even though
+they cost more $$/hour, we spend less amount of time on them.
+
+The problem
+-----------
+
+The thing is that, as of now, every builder in Copr is started by Resalloc "in
+advance", or "preallocated".  The builders are preallocated even though they are
+not _immediately_ used, and stay idle till someone takes them (if there's no
+demand for days, they stay allocated).  When taken, system allocates additional
+resources (so we get more and more machines if there's actual demand for them).
+
+This system works well for normal builders, because they are typically taken
+very quickly (the more tasks in queue the faster they are taken; the smaller the
+"prealloc" quota number for resalloc pool is the faster they are taken).  And
+even if they are not taken for a while, we can actually afford keeping a few of
+them allocated because the instances are not overly expensive.
+
+There's though a demand to have a more (maybe ridiculously) powerful instances;
+having just one of them allocated but unused could be quite expensive matter.
+Even if we used for a low-demanding task.
+
+
+The proposal overview
+---------------------
+
+Let's have on-demand resources/pools in Resalloc.  In such pools there would be
+**no preallocated resource** by default, as long as there's no demand for them.
+Client will have a possibility to ask for them via special "on demand" tagged
+tickets.
+
+We want to be able to
+
+    - allow-list for supported architectures (because for some
+      architectures like s390x it might be quite hard to get "extra_powerful"
+      machines).
```

Comment:
Can this be done with logic "give me VM from pool with tag 'extra_powerfull' if there is no such pool, give me VM from normal pool? Or the other way round, that s390x pool will have only one pool and it will have tag extra_powerfull. So if you spin up VM from normal pool it will go from the same pool as request for extra_powerfull build.

Reply:
Probably doable, but I'm afraid of complications on the Resalloc side; will have to do some magic tickets and tags, like "I can relax the requirement for extra_powerful tag" but not for "arch_x86_64" tag, all this is complicated with the priority queue(s) while scheduling.  And the idea behind the current proposal is already veeery easy to implement and maintain. Let's take a look at this during the meeting.

Reply:
Just a note:  Skipping this, no complications on the Resalloc side for now, nor in Copr -> we'll have a pattern matching mechanism that will not assign extra_powerful tags to the incompatible chroots.